Print module names for "inexhaustive case expression" error

GearsDatapacks opened the pull request:
Closes #3388
The error reporting code now uses the current environment to print module names.

lpil reviewed (commented):
Hello! The aim here is to print variants using the syntax that would be correct for the current module, this instead prints variants with the module name prefixed before them regardless of whether it's valid for the module, or valid Gleam syntax at all.

This is much more involved as it requires the printer to know what is and is not in scope for the current module. Here's an example of something similar for types, though this is not yet hooked up yet to the analyser so that context is never provided. https://github.com/gleam-lang/gleam/blob/7e2d1f247016a9700a4aa9d9e04b59f7c0a6670f/compiler-core/src/type_/printer.rs

GearsDatapacks commented:
Thanks, I'll take a look into it

GearsDatapacks commented:
Question about the implementation of this: Should I add a `print_pattern` method to the `Printer` and integrate it with the analyser? Or should I create a separate `PatternPrinter`?

lpil commented:
A new printer would be good, and they can share the same "names" class, or a different names which could also track values.

GearsDatapacks commented:
I've implemented a `Printer` for missing patterns. I added a way to track defined values to the `TypeNames` class, since I thought it would be easier to just have one struct to keep track of everything.
Although I think we need to wait until #3312 is complete to continue with this, because this needs the context to be provided to it. 
Also I have no idea why the tests are complaining about those unused imports. They aren't unused, they are both used in test functions

GearsDatapacks commented:
Although, since #3312 seems to have been inactive for a while now, maybe it's just worth implementing a separate `ValueNames` struct which just tracks the values for now, to get this working, and then merging with `TypeNames` later, once the other PR is merged.
I guess I'll have a go at that, and see if I can get that working before `TypeNames` is complete

GearsDatapacks commented:
I have now implemented `ValueNames`, which is basically `TypeNames` but simplified for what is needed here, and I've hooked it up to the analyser so it gives correct annotations.

GearsDatapacks commented on compiler-core/src/exhaustiveness/printer.rs:
```diff
@@ -0,0 +1,432 @@
+    // I don't know why the compiler complains about these unused values, they are all
+    // used, but there seems to be a bug where if they're only used in tests,
+    // it doesn't count them as being used. If you're reading this in the future,
+    // try removing this and see if it doesn't complain anymore. (Rust v1.80.0)
```
Ah thank you

GearsDatapacks commented:
With shadowing the prelude, I realised that this applies to any unqualified import which gets shadowed, so I have now fixed that. Good catch!
I've also changed tuple patterns from using a constructor pattern with the name `#`, to have their own variant, because that solution was hacky and it was causing problems.

lpil commented on compiler-core/src/exhaustiveness/printer.rs:
```diff
@@ -0,0 +1,478 @@
+        if let Some(name) = self.local_constructors.get(&key) {
+            // Only return unqualified syntax if the constructor is not shadowed,
+            // and unqualified syntax is valid.
+            eprintln!("{:?}", self.constructor_names.get(name));
```
Debug statement here! If you use `dbg!` clippy will warn you about it

lpil reviewed (approved):
Thank you so much!!!
